File: leetcode/algorithms/704_binary_search/main.java
class BinarySearch {
    public int search(int[] nums, int target) {
        int left = 0;
        int right = nums.length - 1;

        while (left <= right) {
            int mid = left + (right - left) / 2;

            if (nums[mid] < target) {
                left = mid + 1;
            } else if (nums[mid] > target) {
                right = mid - 1;
            } else {
                return mid;
            }
        }

        return -1;
    }


    // Solution
    public int solution(int[] nums, int target) {
        int lo = 0, hi = nums.length - 1;

        while (lo < hi) {
            int mid = lo + (hi - lo + 1) / 2;

            if (target < nums[mid]) {
                hi = mid - 1;
            } else {
                lo = mid;
            }
        }

        return nums[lo] == target ? lo : -1;
    }
}

File: leetcode/algorithms/704_binary_search/main.py
import math
from typing import List


class BinarySearch:
    def search(self, nums: List[int], target: int) -> int:
        left: int = 0
        right: int = len(nums) - 1

        while left <= right:
            mid: int = left + math.floor((right - left) / 2)

            if nums[mid] < target:
                left = mid + 1
            elif nums[mid] > target:
                right = mid - 1
            else:
                return mid

        return -1


    # Solution
    def solution(self, nums: List[int], target: int) -> int:
        lo: int = 0
        hi: int = len(nums) - 1

        while lo < hi:
            mid: int = lo + math.floor((hi - lo + 1) / 2)

            if target < nums[mid]:
                hi = mid - 1
            else:
                lo = mid

        return lo if nums[lo] == target else -1
